_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+
+![activesupport-3.2.2/lib/active_support/inflector/methods.rb:229](/images/activesupport-3-2-2-lib-active-support-inflector-methods-line-229.png)
+
+So, since the first line of the backtrace (`/Users/luke.imhoff/.rvm/gems/ruby-1.9.3-p551@pro/gems/activesupport-3.2.2/lib/active_support/inflector/methods.rb:229:in 'const_defined?'`) mentioned `const_defined?`, I assume that `const_defined?` itself raised the exception and `name` is malformed.  I'm assuming `name` is a `String` (and not a symbol or object that responds to `#to_s`) since it comes from `names` and `names` comes from `camel_cased_word`, which calls `split`, which looks like `String#split`.  `#<KLASS:HEX>` is the MRI format for objects that don't have a an overridden `#inspect` method, so I need to figure out how `constantize` is being called on a `Module` that doesn't define `#inspect`.
+
+`ActiveSupport::Inflector#constantize` is a pure-function (it doesn't use or modify `self`), so I need to just trace `camel_cased_word` back through the call stack.
+
+### `ModuleConstMissing#const_missing`
+
+The next method is  [`ModuleConstMissing#const_missing`](https://github.com/rails/rails/blob/v3.2.2/activesupport/lib/active_support/dependencies.rb#L173-L203).
+
+![`ModuleConstMissing#const_missing`](/images/module-const-missing-const-missing.png)
+
+From line 192, we now know that the `camel_case_word` argument in `ActiveSupport::Inspect#constantize` is the `namespace` local here in `ModuleConstMissing#const_missing`.  `namespace` is an element of [`nesting`](https://github.com/rails/rails/blob/v3.2.2/activesupport/lib/active_support/dependencies.rb#L190).  `nesting` could either be [passed in](https://github.com/rails/rails/blob/v3.2.2/activesupport/lib/active_support/dependencies.rb#L175) or [derived from `klass_name`](https://github.com/rails/rails/blob/v3.2.2/activesupport/lib/active_support/dependencies.rb#L182-L183).  `const_missing` is called implicitly by the Ruby VM since it's not invoked explicitly in [auxiliary/pro/social_engineering/web_phish](https://github.com/rapid7/pro/blob/3370204851ac2d93ba2b3816b3762350799c21df/modules/auxiliary/pro/social_engineering/web_phish.rb#L56)
+
+![`auxiliary/pro/social_engineering/web_phish` `Metasploit3#run`](/images/auxiliary-pro-social-engineering-web-phish-run.png)
+
+This leaves the data flow for `nesting` being `name` → `name.presence` → `klass_name` → `klass_name.to_s` → `klass_name.to_s.scan` → `nesting`.  The original messed up `namespace`, which was `#<Module:0x0000010c916448>` must either be either the direct `Module#name` if `Module#name` contains no `::` or be part of `Module#name`.
```

Comment:
This leaves the data flow for `nesting` being `name` → `name.presence` → `klass_name` → `klass_name.to_s` → `klass_name.to_s.scan` → `nesting`.  The original messed up `namespace`, which was `#<Module:0x0000010c916448>` must **~~either~~** be either the direct `Module#name` if `Module#name` contains no `::` or be part of `Module#name`.